Выполненное ДЗ №4

AntoninaFantalina opened the pull request.

AntoninaFantalina committed:
Добавление файлов ".travis.yml", "Lion.py", "LionTest.py". Изменение файла "README.md"
Файл "Lion.py" содержит конечный автомат, "LionTest.py" тесты для проверки конечного автомата. Файл Travis CI ".travis.yml" для автоматического запуска тестов
В "README.md" 2-й строкой добавлен бейдж Travis CI

AntoninaFantalina committed:
Изменение бейджа Travis CI в файле "README.md"

AntoninaFantalina committed:
Изменение файлов "Lion.py" и "LionTest.py"
Lion.py - добавление проверки правильности состояний в конструкторе, назначение приватными некоторых методов
LionTest.py - разбитие тестов на несколько отдельных с адекватными названиями, добавление негативных тестов

AntoninaFantalina committed:
Изменение бейджа в "README.md" для пулл-реквеста проекта в ветку мастер 2gisprojectT

skyline-gleb commented on Lion.py:
```diff
@@ -0,0 +1,43 @@
+            return None
+
+    def __init__(self, condition):
+        self.state = condition.lower()
```
почему решила использовать **new** и **init** ? сначала сделала не так.

AntoninaFantalina replied:
Не было проверки инициализации состояния, в тестах учла проверку только на правильные значения. В итоге ничего не запрещало создать льва с неправильным состоянием и использовать его дальше, что не есть хорошо.

skyline-gleb replied:
почему для решения проблемы "создания льва с неправильным состоянием" ты решила применить эти две констурукции?  почему не остаться только с init'ом?

AntoninaFantalina replied:
Про проблему льва с неправильным состоянием сегодня у тебя уточнила на лекции, буду делать просто состояние по умолчанию, в случае неправильного задания состояния. Ты сказал можно так) хотя я думала не стоит.
Остаться только с init'ом у меня не получилось - в нем нельзя делать return none, пробовала, да и как-то по другому запретить создание льва.
В последнем коммите
https://github.com/AntoninaFantalina/lesson_4_homework/commit/fa1571e3da45932440bba686dc4c3aa03c9f9897
я сделала вариант что лев создается, но никакое состояние ему не присваивается. Методы в итоге просто отрабатывают в пустую, тоже плохо.
В итоге, как оказалось, ты сказал можно задавать "дефолтное" состояние.

skyline-gleb commented on Lion.py:
```diff
@@ -0,0 +1,43 @@
+    def __init__(self, condition):
+        self.state = condition.lower()
+
+    def transition(self, symbol):
```
почему переименовала `action` -> `transition` ?

AntoninaFantalina replied:
Раньше у класса не было атрибута "действие", оно просто возвращалось. Решила добавить его, мало ли как конечный автомат использовать решат. Поэтому надо переименовать метод (иначе 2 action'а получается). Метод назвала в честь переходов из состояния в состояние по символу.

skyline-gleb replied:
а, я тебя понял, из-за конфликтов имен. ок

skyline-gleb commented on Lion.py:
```diff
@@ -0,0 +1,43 @@
+                self.action = "спать"
+        elif self.state == "сытый":
+                self.state = "голодный"
+                self.action = "смотреть"
```
что-то тут и выше беда с отступами

AntoninaFantalina replied:
Точно. Это у меня сначала вообще весь метод transition (тогда еще action) был единым, без подметодов antelope, hunter, tree, там был немного другой уровень вложенности if'ов. Поправлю.

AntoninaFantalina committed:
Изменение "Lion.py", "LionTest.py"
Возврат от связки  new и init к init, соответствующее изменение негативного теста.
Исправление отступов в Lion.py

AntoninaFantalina committed:
Добавление работы со списками
Изменение логики конечного автомата с добавлением списков состояний/символов/таблиц переходов.

AntoninaFantalina committed:
Изменение "Lion.py"
Добавление инициализации состояния по дефолту при неверном входном состоянии

skyline-gleb commented on Lion.py:
```diff
@@ -0,0 +1,20 @@
+            self.state = state.lower()
+        else:
+            self.state = self.states[0]
+            raise ValueError("Неверное внутреннее состояние ", state)
```
не понимаю этой конструкции. если кидаешь исключение, то зачем делать `elf.state = self.states[0]`? Причем разве не получается, что вернешь всегда исключением что неверное состояние это сытый?
и сразу еще совет - довольно красиво сделать проверку что заданный символ косячный, кинуть исключение, а дальше писать уже код когда все хорошо. так более важный код будет без лишних отступов и не будет лишней конструкции else

AntoninaFantalina replied:
" Причем разве не получается, что вернешь всегда исключением что неверное состояние это сытый?" - да, совершенно точно косяк. И остальные замечания тоже поняла, исправлю.

skyline-gleb commented on Lion.py:
```diff
@@ -0,0 +1,20 @@
+            self.action = self.trans_table_actions[self.symbols.index(symbol)][self.states.index(self.state)]
+            self.state = self.trans_table_states[self.symbols.index(symbol)][self.states.index(self.state)]
+        else:
+            raise ValueError("Неверный входной символ ", symbol)
```
и тут совет мой тоже работает очень даже подходяще

skyline-gleb commented on LionTest.py:
```diff
@@ -0,0 +1,64 @@
+
+    def test_transition_full_negative(self):
+        lion = Lion("сытый")
+        self.assertRaises(ValueError, lion.transition, "НЛО")
```
вообще суть использования списков/таблиц и т.п. еще и в том, что их инициализацию можно вынести через конструктор, тогда в тестах ты вольна задавать свои входные данные, мимнимальные, для полной провеки льва, то есть тестов было бы меньше. оставишь так, ну ок

AntoninaFantalina committed:
Изменение "Lion.py"
Переход от списка к словарю, вынос инициализации словаря через конструктор. Изменение проверок в методах на красивые.

AntoninaFantalina committed:
Изменение "LionTest.py"
Изменение тестов с учетом выноса инициализации словаря в Lion'e, уменьшение количества тестов при минимальных входных данных.

AntoninaFantalina commented:
Не знаю, насколько разумно выносить инициализацию списка/словаря переходов через конструктор. Пользователь конечного автомата может теперь не только накосячить при подаче состояния/символа, но и вовсе задать свою таблицу переходов со своими состояниями и всеми вытекающими, хотя в задании было указано заранее, как эта таблица переходов выглядит. Но, раз говоришь можно, значит можно.

skyline-gleb commented:
> Не знаю, насколько разумно выносить инициализацию списка/словаря переходов через конструктор. 

а зачем тогда сделала? я ведь действительно сказал что можно, а не нужно.
НО:

> Пользователь конечного автомата может теперь не только накосячить при подаче состояния/символа, но и вовсе задать свою таблицу переходов со своими состояниями и всеми вытекающими

так а как с этим бороться, что бы такого не было?

skyline-gleb commented:
и ты сама слишком упростила свой объект `trans_table` из-за чего пришлось накрутить вокруг него дикости в тестах. Эти дикости плохо, что говорит о том что не надо делать невалидные входные данные, надо минимальные

AntoninaFantalina commented:
Восприняла это как совет, потому сделала.

> суть использования списков/таблиц и т.п. еще и в том, что их инициализацию можно вынести через конструктор

По поводу косяков пользователя:

> так а как с этим бороться, что бы такого не было?

Контролировать, что он передал верную таблицу переходов, и после присваивания делать объект trans_table приватным? тогда зачем вообще ее передавать было, если она (в рамках задания) имеет один и тот же вид, сразу задать можно было внутри init'а, без передачи.

AntoninaFantalina committed:
Изменение "LionTest.py"
Устранение "дикости" в тестах, добавление вспомогательных переменных

AntoninaFantalina committed:
Изменение "Lion.py"
Метод lower() применяется к переменным единожды

skyline-gleb commented:
У льва теперь `state = state.lower()` и `symbol = symbol.lower()` делать не правильно, ведь таблицу полученную ты принимаешь как есть, и если там есть большие буквы, ничего не сработает.
Вообщем не надо менять то, что не контролируешь, это создает непонятное поведение.
Однако это уже на заметку и особо не повод не принять :+1: 
p.s. предполагаю что может быть подразумевала входную таблицу в формате обязательно с текстом без заглавныз букв, а формат же не проверям, я сам говорил. вообщем поэтому принимаю в таком виде
